udp_proxy: support file access logging (#20103)
envoyproxy/envoy: docs/root/configuration/observability/access_log/usage.rst

  TCP
    Not implemented (0).

  UDP

>     Number of downstream datagrams with no healthy host found in the upstream cluster. (reused operator for udp proxy).

Reviewer: I'm very sorry if I missed this before, but I think the reuse of this formatter is very confusing. I think we have 2 options that we could potentially use:

1) Add generic new formatters that could cover these cases. If these seem very specific to UDP this is probably not a great option.
2) Add a UDP specific metadata and then allow logging through it via the `DYNAMIC_METADATA` formatter. IMO this is the better option as then you can have very UDP specific things and document that clearly along side the UDP proxy filter.

My recommendation is to go with (2) for the UDP specific things. I think it will end up clean and easy to understand. What do you think?

Author: Yeah, indeed the current docs looks confusing, and (2) looks more convenient and has good compatibility, I will implement it soon.

> I'm very sorry if I missed this before, but I think the reuse of this formatter is very confusing. I think we have 2 options that we could potentially use:
> 
> 1. Add generic new formatters that could cover these cases. If these seem very specific to UDP this is probably not a great option.
> 2. Add a UDP specific metadata and then allow logging through it via the `DYNAMIC_METADATA` formatter. IMO this is the better option as then you can have very UDP specific things and document that clearly along side the UDP proxy filter.
> 
> My recommendation is to go with (2) for the UDP specific things. I think it will end up clean and easy to understand. What do you think?

Yeah, indeed the current docs looks confusing, and (2) looks more convenient and has good compatibility, I will implement it soon.

Author: > I'm very sorry if I missed this before, but I think the reuse of this formatter is very confusing. I think we have 2 options that we could potentially use:
> 
> 1. Add generic new formatters that could cover these cases. If these seem very specific to UDP this is probably not a great option.
> 2. Add a UDP specific metadata and then allow logging through it via the `DYNAMIC_METADATA` formatter. IMO this is the better option as then you can have very UDP specific things and document that clearly along side the UDP proxy filter.
> 
> My recommendation is to go with (2) for the UDP specific things. I think it will end up clean and easy to understand. What do you think?

Yeah, indeed the current docs looks confusing, and (2) looks more convenient and has good compatibility, I will implement it soon.

Reviewer: These docs need cleaning up now.

  TCP
    Not implemented ("-").

  UDP
    For :ref:`UDP Proxy <config_udp_listener_filters_udp_proxy>`,

>     NAMESPACE should be always set to "udp.proxy", optional KEYs are as followed:

Reviewer: ```suggestion
    NAMESPACE should be always set to "udp.proxy", optional KEYs are as follows:
```

  TCP
    Not implemented ("-").

  UDP
    For :ref:`UDP Proxy <config_udp_listener_filters_udp_proxy>`,
    NAMESPACE should be always set to "udp.proxy", optional KEYs are as follows:

    * ``cluster_name``: Name of the cluster.
    * ``bytes_sent``: Total number of downstream bytes sent to the upstream in the session.
    * ``bytes_received``: Total number of downstream bytes received from the upstream in the session.

>     * ``errors_sent``: Number of errors that have occurred When sending datagrams to the upstream in the session.

Reviewer: ```suggestion
    * ``errors_sent``: Number of errors that have occurred when sending datagrams to the upstream in the session.
```

>     * ``errors_received``: Number of errors that have occurred When receiving datagrams from the upstream in UDP proxy.
>       Since the receiving errors are counted in onReceiveError, it seems that no way to distinguish the number of errors in each session.
>       So, it's counted during the lifetime of UDP proxy.

Reviewer: ```suggestion
    * ``errors_received``: Number of errors that have occurred when receiving datagrams from the upstream in UDP proxy.
      Since the receiving errors are counted in at the listener level (vs. the session), this counter is global to all sessions and may not be directly attributable to the session being logged.
```

Author: > 

OK, I have updated docs.Thanks!

Author: > 

OK, I have updated docs.Thanks!
